### ISSUE TITLE
Different results with model.evaluate and model.predict

### ISSUE THREAD

**Issue by simonroquette** [qubvel/segmentation_models#354]
Hello,

I am using your Unet implementation with resnet34 backbone for image segmentation, with encoder frozen. My model predicts 2 classes (True / False) so output shape is (None, 256, 256, 2)

My model's metrics are the following :
- sm.metrics.IOUScore(class_weights=[1, 1], name="iou_avg"), 
- sm.metrics.IOUScore(class_weights=[0, 2], name="iou_true"),
- sm.metrics.IOUScore(class_weights=[2, 0], name="iou_false")

(I use a weight of 2 for IOU score of one class, because I have found that the sum of weights has to be equal to the number of classes to get correct values, but I might be wrong)

The loss function is : sm.losses.JaccardLoss(class_weights=[1, 1])

During training I get the following results : 

Epoch 1/20
8030/8030 [==============================] - 94s 12ms/step - loss: 0.2947 - iou_avg: 0.7053 - iou_true: 0.4343 - iou_false: 0.9764 - val_loss: 0.4595 - val_iou_avg: 0.5405 - val_iou_true: 0.0826 - val_iou_false: 0.9983
Epoch 2/20
8030/8030 [==============================] - 86s 11ms/step - loss: 0.2189 - iou_avg: 0.7811 - iou_true: 0.5694 - iou_false: 0.9928 - val_loss: 0.4708 - val_iou_avg: 0.5292 - val_iou_true: 0.0703 - val_iou_false: 0.9881
Epoch 3/20
8030/8030 [==============================] - 84s 11ms/step - loss: 0.1936 - iou_avg: 0.8064 - iou_true: 0.6188 - iou_false: 0.9939 - val_loss: 0.4372 - val_iou_avg: 0.5628 - val_iou_true: 0.1284 - val_iou_false: 0.9972
Epoch 4/20
8030/8030 [==============================] - 86s 11ms/step - loss: 0.1861 - iou_avg: 0.8140 - iou_true: 0.6337 - iou_false: 0.9942 - val_loss: 0.4389 - val_iou_avg: 0.5611 - val_iou_true: 0.1253 - val_iou_false: 0.9970
... (~constant results)
Epoch 9/20
8030/8030 [==============================] - 86s 11ms/step - loss: 0.1326 - iou_avg: 0.8674 - iou_true: 0.7386 - iou_false: 0.9962 - val_loss: 0.4333 - val_iou_avg: 0.5667 - val_iou_true: 0.1355 - val_iou_false: 0.9978
Epoch 10/20
8030/8030 [==============================] - 85s 11ms/step - loss: 0.1274 - iou_avg: 0.8726 - iou_true: 0.7490 - iou_false: 0.9963 - val_loss: 0.4337 - val_iou_avg: 0.5663 - val_iou_true: 0.1351 - val_iou_false: 0.9975
Restoring model weights from the end of the best epoch
Epoch 00010: early stopping

Training metrics look promising, but validation don't. One might say this could be overfitting, but when I make predictions on my test set after training, I get really good results, that also look good when I visualize my images.

When I use model.evaluate(X_test, Y_test) I get : 
 [0.405 (loss), 0.585 (iou_avg), 0.220 (iou_true), 0.950 (iou_false)]

When manually computing the loss/iou score using sm.metrics.IOUScore(class_weights=[1, 1])(model.predict(X_test), Y_test)  (changing the weights for each value) I get much better results : 
[0.199 (loss), 0.801 (iou_avg), 0.652 (iou_true), 0.950 (iou_false)]

Following existing posts about similar issues, I have considered the possibility that the problem came from Batch Normalization layer which can have problems if you do transfer learning while freezing some layers ( http://blog.datumbox.com/the-batch-normalization-layer-of-keras-is-broken/). I tried forcing the batch normalization layers trainable parameter to False, and it didn't help.

Thank you if you in advance for your suggestions and help.

**Comment by silvestrebahi** [comment]
Hi! From what I see here you just ran the model on 10 epochs. Also I don't know what kind of dataset you are using, but your batch size looks small (from the 11ms/step). If your model contains BatchNormalization layers this looks pretty normal, as they are implemented to use the current batch mean and std during training and then use the whole dataset's mean and std for inference. If your model contains this kind of layer, I'd suggest to train for longer with a bigger batch size.
